### PR TITLE
Add auth0

### PR DESCRIPTION
**Pull request by landonreed**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Depends on #12. This PR adds authentication via Auth0 (much of this code is taken from ibi-group/datatools-server). It also adds a UserController class that uses hooks into ApiController in order to inject some validation of users and create an actual Auth0 user when creating users in the otp-middleware application. Note: there are updates to env.yml for Auth0 settings.

Works with ibi-group/otp-admin-ui#1

### CODE REVIEW

File: pom.xml
Comment (binh-dam-ibigroup) [inline pom.xml]: JDK 11 ships with its own http client class `java.net.http.HttpClient`.

File: src/main/java/org/opentripplanner/middleware/auth/Auth0Users.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/auth/Auth0Users.java]: Try using `java.net.http.HttpClient` and related classes instead.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
I have more to review but I wanted to get this one out to use the built-in `java.net.http.HttpClient` and related classes.

File: src/main/java/org/opentripplanner/middleware/auth/Auth0Users.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/auth/Auth0Users.java]: Will these subscriptions be a thing in FDOT? If not, let's remove. We should also move away from storing user data in the auth0 app_metadata as well.

**Review by evansiroky — COMMENTED** [review]
Is there a UI this can be tested with? Also, it seems a lot of this is copied from datatools. Any way we can copy over some of the tests in there too?

**Comment by landonreed** [comment]
@evansiroky, you can test this out with https://otp-admin-ui.now.sh/ (technically it's running the `aws-sdk-integration` branch, but all of the user/Auth0 stuff is there, too).

Yes, probably should copy over the tests and remove the unused methods.

File: src/main/java/org/opentripplanner/middleware/auth/Auth0Users.java
Comment (landonreed) [reply src/main/java/org/opentripplanner/middleware/auth/Auth0Users.java]: Notifications will definitely be a thing. But we won't be storing things in app_metadata. I think we should just plan to use the user collection for this stuff. We need to figure out how we want to store various user types (users of OTP vs. users of admin dashboard).

File: pom.xml
Comment (landonreed) [reply pom.xml]: I tried using the java `HttpClient`, but can't necessarily get rid of this package because it provides `UriBuilder`. I don't think there's a great alternative to that class yet.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Is the intent to make the pre-execution checks mandatory?

File: src/main/java/org/opentripplanner/middleware/models/User.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/User.java]: By storing both `openTripPlannerOptions` and `adminOptions` in the same class, it looks like we are mashing together the users of OTP and the users of the admin dashboard and saying that a regular OTP user could have access to the admin dashboard... which one is it? I think both kinds of users should be separate (they can derive from a common user class though).

**Review by binh-dam-ibigroup — DISMISSED** [review]
With the changes made, I'm inclined to approve assuming we are planning to deal with user types.

**Comment by landonreed** [comment]
@evansiroky, any thoughts re testing this new code (it uses the new auth0 java client instead of a bunch of the old Data Tools code)?

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Also prevent changes to other Auth0 users unless an admin?

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: There should be a check for whether the user making the request is authorized to delete the other user.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See 2 comments about having some authorization checks for updating and deleting user info.

Also, it seems like we could bring over the `canCreateUser` and `canDeleteUser` tests from com.conveyal.datatools.manager.controller.api.UserControllerTest without significant code changes.

**Comment by landonreed** [comment]
Thanks @evansiroky, also I just realized that I had some stashed changes that I need to add here, so there might be a few additional changes inbound shortly.

**Comment by evansiroky** [comment]
Unassigning myself until #14 is merged in here.

File: src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java]: Typo `user is ...`

**Review by binh-dam-ibigroup — DISMISSED** [review]
Minor typo but good to go as is.

**Review by binh-dam-ibigroup — DISMISSED** [review]

**Comment by evansiroky** [comment]
Unassigning myself until #14 is merged in here.

**Review by evansiroky — APPROVED** [review]
Seems like everything was reviewed in #14.

**Comment by landonreed** [comment]
Yes, agreed, @evansiroky. I'm going to go ahead and merge.

**Pull request merged**
